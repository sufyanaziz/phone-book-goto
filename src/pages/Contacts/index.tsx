import React, { useCallback, useContext, useMemo } from "react";
import Loader from "@common/components/Loader";
import Button from "@common/components/Button";
import useContact from "./hooks/useContact";
import CardContact from "./components/CardContact";
import Text from "@common/components/Text";
import Container from "@common/components/Container";
import HeaderTitle from "@common/components/HeaderTitle";
import SearchBar from "./components/SearchBar";
import { ContactStore } from "@common/store/useContactStore";
import "./style/index.css";

const LIMIT = 10;

const Contacts = () => {
  const { offset, onSetOffset } = useContext(ContactStore);
  const { data, loading, totalRow } = useContact({ limit: LIMIT });

  const pagination = useMemo(() => {
    return {
      totalList: data?.length || 0,
      totalRow,
      currentPage: (LIMIT + offset) / LIMIT,
      totalPage: Math.ceil(totalRow / LIMIT),
      activePage: `${(LIMIT + offset) / LIMIT}/${Math.ceil(totalRow / LIMIT)}`,
    };
  }, [data, offset, totalRow]);

  const renderData = useCallback(() => {
    if (loading) return <Loader />;
    if (!data?.length) return <Text text="Data is not found" />;
    return data.map((item) => {
      return <CardContact key={item.id} data={item} />;
    });
  }, [data, loading]);

  return (
    <div className="contact-page">
      <Container>
        <HeaderTitle title="Contact List" />
        <SearchBar />
        <div className="contact-page-content" css={{ marginBottom: 12 }}>
          {renderData()}
        </div>
        <div
          css={{
            display: "flex",
            alignItems: "center",
            justifyContent: "space-between",
          }}
        >
          <Text
            text={`showing ${pagination.totalList} of ${totalRow} data (${pagination.activePage})`}
          />
          <div css={{ display: "flex", gap: "4px" }}>
            <Button
              label="Prev"
              disabled={!offset}
              onClick={() => {
                onSetOffset(offset - LIMIT);
              }}
            />
            <Button
              label="Next"
              disabled={
                data?.length !== LIMIT || !((LIMIT + offset) % totalRow)
              }
              onClick={() => {
                onSetOffset(offset + LIMIT);
              }}
            />
          </div>
        </div>
      </Container>
    </div>
  );
};

export default Contacts;
