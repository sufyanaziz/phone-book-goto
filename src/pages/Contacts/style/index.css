.contact-page {
  height: calc(100vh - 64px);
}

.contact-page-content {
  height: 400px;
  overflow-y: auto;
}
