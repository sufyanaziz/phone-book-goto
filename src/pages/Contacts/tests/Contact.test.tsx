import React from "react";
import renderer from "react-test-renderer";
import { BrowserRouter as Router } from "react-router-dom";
import { ApolloClient, ApolloProvider, InMemoryCache } from "@apollo/client";
import Contacts from "@pages/Contacts";

const client = new ApolloClient({
  uri: "",
  cache: new InMemoryCache(),
});

describe("Component Contacts", () => {
  it("should render Contacts component correctly", () => {
    const component = renderer.create(
      <ApolloProvider client={client}>
        <Router>
          <Contacts />
        </Router>
      </ApolloProvider>
    );
    expect(component).toMatchSnapshot();
  });
});
