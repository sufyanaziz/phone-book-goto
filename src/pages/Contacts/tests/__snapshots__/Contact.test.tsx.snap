// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Contacts should render Contacts component correctly 1`] = `
<div
  className="contact-page"
>
  <div
    css={[Function]}
  >
    <div
      css={[Function]}
    >
      <div
        className=""
        css={[Function]}
      >
        Contact List
      </div>
    </div>
    <div
      css={[Function]}
    >
      <input
        className="input"
        onChange={[Function]}
        placeholder="Find Contact"
        value=""
      />
      <div>
        <button
          className="primary"
          css={[Function]}
          onClick={[Function]}
        >
          Add New Contact
        </button>
      </div>
    </div>
    <div
      className="contact-page-content"
      css={
        {
          "marginBottom": 12,
        }
      }
    >
      <div
        className=""
        css={[Function]}
      >
        loading...
      </div>
    </div>
    <div
      css={
        {
          "alignItems": "center",
          "display": "flex",
          "justifyContent": "space-between",
        }
      }
    >
      <div
        className=""
        css={[Function]}
      >
        showing 0 of 0 data (1/0)
      </div>
      <div
        css={
          {
            "display": "flex",
            "gap": "4px",
          }
        }
      >
        <button
          className="primary"
          css={[Function]}
          disabled={true}
          onClick={[Function]}
        >
          Prev
        </button>
        <button
          className="primary"
          css={[Function]}
          disabled={true}
          onClick={[Function]}
        >
          Next
        </button>
      </div>
    </div>
  </div>
  <div
    css={[Function]}
  >
    <div
      className="modal-mask"
      onClick={[Function]}
    />
    <div
      className="modal-content"
    >
      <div
        className="header"
      >
        <div
          css={
            {
              "alignItems": "center",
              "display": "flex",
            }
          }
        >
          <div
            className="header-title"
            css={[Function]}
          >
            Detail Contact
          </div>
          <div
            css={
              {
                "marginLeft": "8px",
              }
            }
          >
            <button
              className="warning"
              css={[Function]}
              onClick={[Function]}
            >
              Edit
            </button>
          </div>
        </div>
        <div
          className="close-icon"
          onClick={[Function]}
        >
          <div
            className=""
            css={[Function]}
          >
            X
          </div>
        </div>
      </div>
      <div
        className="content"
      >
        <div
          css={[Function]}
        >
          <div
            className="section-name"
          >
            <div
              className="first-name"
            >
              <div
                className="header-title"
                css={[Function]}
              >
                First Name
              </div>
              <div
                className=""
                css={[Function]}
              >
                -
              </div>
            </div>
            <div
              className="last-name"
            >
              <div
                className="header-title"
                css={[Function]}
              >
                Last Name
              </div>
              <div
                className=""
                css={[Function]}
              >
                -
              </div>
            </div>
          </div>
          <div
            className="section-phone-number"
          >
            <div
              className="header-title"
              css={[Function]}
            >
              Phones
            </div>
            -
          </div>
        </div>
      </div>
      <div
        className="footer"
      />
    </div>
  </div>
  <div
    css={[Function]}
  >
    <div
      className="modal-mask"
      onClick={[Function]}
    />
    <div
      className="modal-content"
    >
      <div
        className="header"
      >
        <div
          className=""
          css={[Function]}
        >
          Add New Contact
        </div>
        <div
          className="close-icon"
          onClick={[Function]}
        >
          <div
            className=""
            css={[Function]}
          >
            X
          </div>
        </div>
      </div>
      <div
        className="content"
      >
        <div
          css={[Function]}
        >
          <div
            className="section-name"
          >
            <div
              className="first-name"
            >
              <div
                className="header-title"
                css={[Function]}
              >
                First Name
              </div>
              <input
                className="input"
                disabled={false}
                onChange={[Function]}
                placeholder="Input First Name..."
                value=""
              />
            </div>
            <div
              className="last-name"
            >
              <div
                className="header-title"
                css={[Function]}
              >
                Last Name
              </div>
              <input
                className="input"
                disabled={false}
                onChange={[Function]}
                placeholder="Input Last Name..."
                value=""
              />
            </div>
          </div>
          <div
            className="section-phone-number"
          >
            <div
              className="header-title"
              css={[Function]}
            >
              Phones
            </div>
            <div
              className="phone-container"
            >
              <input
                className="input"
                disabled={false}
                onChange={[Function]}
                placeholder="Phone 1"
                value=""
              />
              <div
                className="icons"
              >
                <div
                  className="add-icon"
                  onClick={[Function]}
                >
                  <svg
                    fill="none"
                    height="24px"
                    viewBox="0 0 24 24"
                    width="23px"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M9 12H15"
                      stroke="#435334"
                      strokeLinecap="round"
                      strokeLinejoin="round"
                      strokeWidth="2"
                    />
                    <path
                      d="M12 9L12 15"
                      stroke="#435334"
                      strokeLinecap="round"
                      strokeLinejoin="round"
                      strokeWidth="2"
                    />
                    <path
                      d="M3 12C3 4.5885 4.5885 3 12 3C19.4115 3 21 4.5885 21 12C21 19.4115 19.4115 21 12 21C4.5885 21 3 19.4115 3 12Z"
                      stroke="#435334"
                      strokeWidth="2"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="footer"
      >
        <div
          className="flex"
          css={
            {
              "padding": "0 12px 12px 12px",
            }
          }
        >
          <button
            className="primary"
            css={[Function]}
            disabled={true}
            onClick={[Function]}
          >
            Submit
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    css={[Function]}
  >
    <div
      className="modal-mask"
      onClick={[Function]}
    />
    <div
      className="modal-content"
    >
      <div
        className="header"
      >
        <div
          className=""
          css={[Function]}
        >
          Edit Contact
        </div>
        <div
          className="close-icon"
          onClick={[Function]}
        >
          <div
            className=""
            css={[Function]}
          >
            X
          </div>
        </div>
      </div>
      <div
        className="content"
      >
        <div
          css={[Function]}
        >
          <div
            className="section-name"
          >
            <div
              className="first-name"
            >
              <div
                className="header-title"
                css={[Function]}
              >
                First Name
              </div>
              <input
                className="input"
                disabled={false}
                onChange={[Function]}
                placeholder="Input First Name..."
                value=""
              />
            </div>
            <div
              className="last-name"
            >
              <div
                className="header-title"
                css={[Function]}
              >
                Last Name
              </div>
              <input
                className="input"
                disabled={false}
                onChange={[Function]}
                placeholder="Input Last Name..."
                value=""
              />
            </div>
          </div>
          <div
            className="section-phone-number"
          >
            <div
              className="header-title"
              css={[Function]}
            >
              Phones
            </div>
            <div
              className="phone-container"
            >
              <input
                className="input"
                disabled={true}
                onChange={[Function]}
                placeholder="Phone 1"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="footer"
      >
        <div
          className="flex"
          css={
            {
              "padding": "0 12px 12px 12px",
            }
          }
        >
          <button
            className="primary"
            css={[Function]}
            disabled={true}
            onClick={[Function]}
          >
            Submit
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
